pkg/command/context.go

```diff
@@ -1673,28 +1673,77 @@ func validateActiveResourceOptions() error {
 	return nil
 }
 
-func updateTanzuContextKubeconfig(cliContext *configtypes.Context, project, spaceName, clustergroupName string) error {
+func updateTanzuContextKubeconfig(cliContext *configtypes.Context, projectName, projectID, spaceName, clustergroupName string) error {
 	if cliContext.ClusterOpts == nil {
-		return errors.New("invalid context. Kubeconfig details missing in the context")
+		return errors.New("invalid context. Kubeconfig details are missing in the context")
 	}
+
 	kcfg, err := clientcmd.LoadFromFile(cliContext.ClusterOpts.Path)
 	if err != nil {
 		return errors.Wrap(err, "unable to load kubeconfig")
 	}
 
-	kubeContext := kcfg.Contexts[cliContext.ClusterOpts.Context]
-	if kubeContext == nil {
-		return errors.Errorf("kubecontext %q doesn't exist", cliContext.ClusterOpts.Context)
+	kubeCtx, cluster, err := getKubeContextAndCluster(kcfg, cliContext)
+	if err != nil {
+		return err
 	}
-	cluster := kcfg.Clusters[kubeContext.Cluster]
-	cluster.Server = prepareClusterServerURL(cliContext, project, spaceName, clustergroupName)
+
+	projectVal := getProjectValueForKubeconfig(projectName, projectID)
+	newServerURL := prepareClusterServerURL(cliContext, projectVal, spaceName, clustergroupName)
+
+	// Manage context based on environment variable
+	useStableKubeContextName, _ := strconv.ParseBool(os.Getenv(constants.UseStableKubeContextNameForTanzuContext))
+	if useStableKubeContextName {
+		updateContextInPlace(cluster, newServerURL)
+	} else {
+		newContextName := prepareKubeContextName(cliContext, projectName, spaceName, clustergroupName)
+		updateContextWithNewName(kcfg, cliContext, cluster, kubeCtx, newContextName, newServerURL)
```

Comment:
nit: It took a lot of time for me to understand what was happening here. This may be because of additional `kubeCtx` and `cluster` variables that we are precalculating.

I get why we are precalculating it so I am fine if you want to keep it the way it is (May be just rename `cluster` to `kubecluster` or something like that).

But maybe another option is to not pre-calculate `kubeCtx` and `cluster` and get that when needed as part of `updateContextInPlace` and `updateContextWithNewName`. Because we are anyways passing the `kcfg` to `updateContextWithNewName` function. So no need to actually pass `cluster` again.

Also, if we just pass `kcfg` to both `updateContextInPlace` and `updateContextWithNewName` both function will look more consistent and easy to understand. This is just my thinking.

Reply:
The idea is to fetch the kubecontext and kube cluster and error out before calling `updateContextInPlace` and `updateContextWithNewName`. I updated to fetch the the kubecontext and kubecluster inside the above functions to avoid confusion that it may bring.